examples_tests/19.SIMDmatrixMultiplication/main.cpp

```diff
@@ -0,0 +1,238 @@
+		{
+			matrix4x3_row out;
+
+			out.row[0] = doJob(row[0], _other);
+			out.row[1] = doJob(row[1], _other);
+			out.row[2] = doJob(row[2], _other);
+			out.row[3] = row[3]; // 0 0 0 1
+
+			return out;
+		}
+
+		float& operator()(size_t _i, size_t _j) { return m[_i][_j]; }
+		const float& operator()(size_t _i, size_t _j) const { return m[_i][_j]; }
+
+	private:
+		static inline __m128 doJob(const __m128& a, const matrix4x3_row& _mtx)
+		{
+			__m128 res;
+			res = _mm_mul_ps(_mm_shuffle_ps(a, a, BROADCAST32(0)), _mtx.row[0]);
+			res = _mm_add_ps(res, _mm_mul_ps(_mm_shuffle_ps(a, a, BROADCAST32(1)), _mtx.row[1]));
+			res = _mm_add_ps(res, _mm_mul_ps(_mm_shuffle_ps(a, a, BROADCAST32(2)), _mtx.row[2]));
+			res = _mm_add_ps(res, _mm_mul_ps(_mm_shuffle_ps(a, a, BROADCAST32(3)), _mtx.row[3]));
+			return res;
+		}
+	}
+#ifndef _IRR_WINDOWS_
+	__attribute__((__aligned__(SIMD_ALIGNMENT)));
+#endif
+	; // matrix4x3_row
+
+#ifdef _IRR_WINDOWS_
+	__declspec(align(SIMD_ALIGNMENT))
+#endif
+	struct matrix4x3_col
+	{
+		union
+		{
+			float m[4][4];
+			__m128 col[4];
+		};
+
+		inline matrix4x3_col concatenate(const matrix4x3_col& _other)
+		{
+			matrix4x3_col out;
+
+			out.col[0] = doJob(_other.col[0], 0, *this);
+			out.col[1] = doJob(_other.col[1], 1, *this);
+			out.col[2] = doJob(_other.col[2], 2, *this);
+			out.col[3] = doJob(_other.col[3], 3, *this);
+
+			return out;
+		}
+
+		float& operator()(size_t _i, size_t _j) { return m[_j][_i]; }
+		const float& operator()(size_t _i, size_t _j) const { return m[_j][_i]; }
+
+	private:
+		static inline __m128 doJob(__m128 a, size_t j, const matrix4x3_col& _mtx)
+		{
+			__m128 res;
+			res = _mm_mul_ps(_mm_shuffle_ps(a, a, BROADCAST32(0)), _mtx.col[0]);
+			res = _mm_add_ps(res, _mm_mul_ps(_mm_shuffle_ps(a, a, BROADCAST32(1)), _mtx.col[1]));
+			res = _mm_add_ps(res, _mm_mul_ps(_mm_shuffle_ps(a, a, BROADCAST32(2)), _mtx.col[2]));
+			if (j == 3) // w of every column except last one is 0; in the last one it's 1
+				res = _mm_add_ps(res, _mtx.col[3]);
+			return res;
+		}
+	}
+#ifndef _IRR_WINDOWS_
+	__attribute__((__aligned__(SIMD_ALIGNMENT)));
+#endif
+	; // matrix4x3_col
+} // sse3
+
+namespace measure
+{
+	using Clock = std::chrono::high_resolution_clock;
+	using TimePoint = std::chrono::time_point<Clock>;
+	using Duration = std::chrono::duration<double, std::micro>;
+}
+
+template<typename T>
+static double run();
+
+int main()
+{
+	for (size_t i = 0; i < 100; ++i)
+	{
+		printf("avx  row: %f\n", run<avx::matrix4x3_row>());
+		printf("avx  col: %f\n", run<avx::matrix4x3_col>());
+		printf("sse3 row: %f\n", run<sse3::matrix4x3_row>());
+		printf("sse3 col: %f\n", run<sse3::matrix4x3_col>());
+	}
+
+	return 0;
+}
+
+template<typename T>
+static double run()
+{
+	float content[16];
+
+	const measure::TimePoint start = measure::Clock::now();
+
+	for (size_t i = 0; i < (size_t)EXEC_CNT; ++i)
+	{
+		for (size_t k = 0; k < 16; ++k)
+			content[k] = rand();
```

Comment:
generate the matrices to be multiplied BEFORE starting the main loop, otherwise rand() will take most of the computation time!